src/Networking/real_world_interface.cpp

```diff
@@ -25,40 +25,89 @@ extern "C"
 
 using nlohmann::json;
 
+struct timeval last_odom_reading_;
+transform_t last_odom_tf_;
+double odom_dtheta_, odom_dx_;
+
 void world_interface_init() {
 	bool gps_success = startGPSThread();
 	bool lidar_success = lidar::initializeLidar();
 	bool landmark_success = AR::initializeLandmarkDetection();
+
+	gettimeofday(&last_odom_reading_, NULL);
+	last_odom_tf_ = toTransform({0., 0., 0.});
+	odom_dtheta_ = 0.0;
+	odom_dx_ = 0.0;
 }
 
-constexpr double WHEEL_BASE = 1.0;			   // Distance between left and right wheels. Eyeballed
-constexpr double WHEEL_RADIUS = 0.15;		   // Eyeballed
-constexpr double PWM_FOR_1RAD_PER_SEC = 10000; // Eyeballed
+transform_t getOdomAt(const struct timeval &time) {
+	double elapsed = getElapsedUsecs(last_odom_reading_, time) / 1000000.0;
+	double delta_theta = odom_dtheta_ * elapsed;
+	double rel_x, rel_y, rel_th;
+	rel_th = odom_dtheta_ * elapsed;
+	if (abs(odom_dtheta_) < 0.000001) {
+		rel_x = odom_dx_ * elapsed;
+		rel_y = 0.0;
+	} else {
+		double turn_radius = odom_dx_ / odom_dtheta_;
+		rel_x = turn_radius * sin(rel_th);
+		rel_y = turn_radius * (1 - cos(rel_th));
+	}
+	transform_t rel_tf = toTransform({rel_x, rel_y, rel_th});
+	return rel_tf * last_odom_tf_;
+}
+
+void setCmdVelToIntegrate(double dtheta, double dx) {
+	struct timeval now;
+	gettimeofday(&now, NULL);
+	transform_t new_tf = getOdomAt(now);
+	last_odom_reading_ = now;
+	last_odom_tf_ = new_tf;
+	odom_dtheta_ = dtheta;
+	odom_dx_ = dx;
+}
+
+// Distance between left and right wheels.
+constexpr double WHEEL_BASE = 0.66;
+// Effective distance between wheels. Tweaked so that actual rover angular rate
+// roughly matches the commanded angular rate.
+constexpr double EFF_WHEEL_BASE = 1.40;
+
+constexpr double WHEEL_RADIUS = 0.15;			 // Eyeballed
+constexpr double PWM_FOR_1RAD_PER_SEC = 5000; // Eyeballed
+// This is a bit on the conservative side, but we heard an ominous popping sound at 20000.
+constexpr double MAX_PWM = 20000;
 
 bool setCmdVel(double dtheta, double dx) {
 	if (Globals::E_STOP && (dtheta != 0 || dx != 0))
 		return false;
 
 	/* This is the inverse of the formula:
-	 *    dx = (right_ground_vel + left_ground_vel) / 2
-	 *    dtheta = (right_ground_vel - left_ground_vel) / WHEEL_BASE
+	 *		dx = (right_ground_vel + left_ground_vel) / 2
+	 *		dtheta = (right_ground_vel - left_ground_vel) / EFF_WHEEL_BASE
 	 */
-	double right_ground_vel = dx + WHEEL_BASE / 2 * dtheta;
-	double left_ground_vel = dx - WHEEL_BASE / 2 * dtheta;
+	double right_ground_vel = dx + EFF_WHEEL_BASE / 2 * dtheta;
```

Comment:
This is pretty minor since the calculations are pretty short, but if you want you could reuse the existing `DiffDriveKinematics` class to do these kind of calculations

Reply:
Yeahhh I was trying to remember where that code was. I figured it would be good practice to re-derive it, but from a software engineering standpoint we probably ought to use just one implementation